cloud/store/client.go

```diff
@@ -30,63 +35,114 @@ func NewClient(localService Cluster) *Client {
+	addr, err := cl.rpcAddressFromRAFT(rAddr)
+	if err != nil {
+		return nil, err
+	}
+	if addr == cl.localService.address {
+		return cl.localService.NotifyPeer(ctx, req)
+	}
+
+	conn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
 	if err != nil {
-		return err
+		return nil, err
 	}
 	defer conn.Close()
 	c := cmd.NewClusterServiceClient(conn)
 	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
 	defer cancel()
-	r, err := c.JoinPeer(ctx, req)
+	return c.NotifyPeer(ctx, req)
+}
+
+func (cl *Client) join(ctx context.Context, leaderAddr string, req *cmd.JoinPeerRequest) (*cmd.JoinPeerResponse, error) {
+	log.Printf("client join: %s %+v\n", leaderAddr, req)
+	addr, err := cl.rpcAddressFromRAFT(leaderAddr)
 	if err != nil {
-		return err
+		return nil, err
+	}
+	if addr == cl.localService.address {
+		return cl.localService.JoinPeer(ctx, req)
 	}
-	log.Printf("join response: %s", r.String())
-	return nil
+
+	conn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+	c := cmd.NewClusterServiceClient(conn)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	return c.JoinPeer(ctx, req)
 }
 
-func (cl Client) Remove(leaderAddress string, req *cmd.RemovePeerRequest) error {
+func (cl *Client) remove(ctx context.Context, leaderAddress string, req *cmd.RemovePeerRequest) (*cmd.RemovePeerResponse, error) {
 	log.Printf("client remove: %s %+v\n", leaderAddress, req)
-	if cl.localService.address == leaderAddress {
-		return cl.localService.Remove(req.Id)
+	addr, err := cl.rpcAddressFromRAFT(leaderAddress)
+	if err != nil {
+		return nil, err
+	}
+	if cl.localService.address == addr {
+		return cl.localService.RemovePeer(ctx, req)
 	}
-	// lAddr := cl.leaderAddress()
-	// if lAddr == "" {
-	// 	return err
-	// }
-	conn, err := grpc.Dial(leaderAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
+
+	conn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
 	if err != nil {
-		return err
+		return nil, err
 	}
 	defer conn.Close()
 	c := cmd.NewClusterServiceClient(conn)
 	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
 	defer cancel()
-	r, err := c.RemovePeer(ctx, req)
-	if err != nil {
-		return err
-	}
-	log.Printf("join response: %s", r.String())
-	return nil
+	return c.RemovePeer(ctx, req)
 }
 
-// TODO-RAFT
-// leaderAddress is just a workaround to get address of leader rpc  service
-func (cl Client) leaderAddress() string {
-	lAddr, _ := cl.localService.Raft.LeaderWithID()
-	if lAddr == "" {
-		return ""
+// rpcAddressFromRAFT returns the RPC address based on the provided RAFT address.
+// In a real cluster, the RPC port is the same for all nodes.
+// In a local environment, the RAFT ports need to be different. Specifically,
+// we calculate the RPC port as the RAFT port + 1.
```

Comment:
We have no guarantee that iPort+1 is available/free on the machine where weaviate is running. We should either rely on default or config, but not pick "the next one".

Reply:
I intended it as a workaround to address the challenges associated with handling different ports for the same host. Feel free to enhance it further to ensure it accommodates edge cases.